File: apps/typescript/design-pattern/src/strategy.ts
/**
 * Strategy パターン
 *
 * アルゴリズムをオブジェクト（関数）として切り出し、
 * 実行時に差し替え可能にする。
 */

export type Formatter = (title: string, text: string[]) => string;

export const htmlFormatter: Formatter = (title, text) => {
  const lines = text.map((line) => `<p>${line}</p>`).join('\n');
  return `<html>\n<head><title>${title}</title></head>\n<body>\n${lines}\n</body>\n</html>`;
};

export const plainTextFormatter: Formatter = (title, text) => {
  const lines = text.map((line) => line).join('\n');
  return `***** ${title} *****\n${lines}`;
};

export class Report {
  private title: string;
  private text: string[];
  private formatter: Formatter;

  constructor(title: string, text: string[], formatter: Formatter) {
    this.title = title;
    this.text = text;
    this.formatter = formatter;
  }

  setFormatter(formatter: Formatter): void {
    this.formatter = formatter;
  }

  outputReport(): string {
    return this.formatter(this.title, this.text);
  }
}

File: apps/python/design-pattern/src/strategy.py
"""Strategy パターン

フォーマット戦略を callable（関数/ラムダ）として受け取り、
出力処理を委譲する。
"""

from typing import Callable, Protocol


class ReportData(Protocol):
    """レポートデータのプロトコル"""

    @property
    def title(self) -> str: ...

    @property
    def text(self) -> list[str]: ...


class Report:
    """レポートクラス（Strategy パターン）"""

    def __init__(self, formatter: "Formatter") -> None:
        self._title: str = "月次報告"
        self._text: list[str] = ["順調", "最高の調子"]
        self.formatter = formatter

    @property
    def title(self) -> str:
        return self._title

    @property
    def text(self) -> list[str]:
        return list(self._text)

    def output_report(self) -> str:
        return self.formatter(self)


# 型エイリアス
Formatter = Callable[["Report"], str]


def html_formatter(report: Report) -> str:
    """HTML フォーマット戦略"""
    lines = [
        "<html>",
        f"  <head><title>{report.title}</title></head>",
        "  <body>",
    ]
    for line in report.text:
        lines.append(f"    <p>{line}</p>")
    lines.append("  </body>")
    lines.append("</html>")
    return "\n".join(lines) + "\n"


def plain_text_formatter(report: Report) -> str:
    """プレーンテキストフォーマット戦略"""
    lines = [f"***** {report.title} *****"]
    for line in report.text:
        lines.append(line)
    return "\n".join(lines) + "\n"
